Memory pair game

Tedzury opened the pull request:
### Memory pair game

[Demo](https://tedzury.github.io/Memory_pair_game/) |
[Code base](https://github.com/Tedzury/Memory_pair_game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

Tedzury commented:
Nooooo, bot, I'm not stale, I'm still here !

Tedzury commented:
I'm still here...

Tedzury commented:
UAT - done. Feel myself ready for review.

A-Ostrovnyy commented on submissions/Oleksii_Drohachov/Memory_pair_game/script.js:
```diff
@@ -0,0 +1,220 @@
+
+function prepareGame() {
+  const cardsArray = [
+    {
+      name: "beast",
+      path: "img/Beast.jpg",
+    },
+    {
+      name: "beast",
+      path: "img/Beast.jpg",
+    },
```
No need to duplicate card items

A-Ostrovnyy commented on submissions/Oleksii_Drohachov/Memory_pair_game/script.js:
```diff
@@ -0,0 +1,220 @@
+      path: "img/Metal.jpg",
+    },
+  ];
+  let shuffled = [...cardsArray];
```
```suggestion
  let shuffled = [...cardsArray, ...cardsArray];
```
Also, if you don't reassign it use const instead.

A-Ostrovnyy commented on submissions/Oleksii_Drohachov/Memory_pair_game/script.js:
```diff
@@ -0,0 +1,220 @@
+
+  function cardsHandler() {
+    const cards = document.querySelectorAll(".card");
+    cards.forEach((card) => card.addEventListener("click", flipCard));
```
User event delegation. [Read more](https://davidwalsh.name/event-delegation)

A-Ostrovnyy commented on submissions/Oleksii_Drohachov/Memory_pair_game/script.js:
```diff
@@ -0,0 +1,220 @@
+      if (lockBoard) return;
+      if (this === firstCard) return;
+
+      this.classList.add("flip");
```
You can easily lose context and the corresponding value of `this` pass selected card as a parameter for the function.

A-Ostrovnyy reviewed (changes requested):
Good work!

Tedzury commented:
Good day. Fixed everuthing you mentioned, applied Event delegation to gameboard, so slightly reefactored game logic, some tricks with event targets. Also refactored modal rendering, put out making of modal from function, in func changing only modal text and button title.

A-Ostrovnyy commented on submissions/Oleksii_Drohachov/Memory_pair_game/script.js:
```diff
@@ -0,0 +1,203 @@
+  function cardsHandler() {
+    gameboard.addEventListener("click", ({ target }) => {
+      if (
+        target.parentElement.classList.contains("card") &&
```
Don't use constructions like children[0], firstElementChild, parentNode, nextSibling, etc. In such way, you rely on the order of DOM elements. So in case when you will have changed design - your code will be broken. Which is bad, obviously. Use querySelector or closest, if in event, instead. [Read more](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)

A-Ostrovnyy commented on submissions/Oleksii_Drohachov/Memory_pair_game/script.js:
```diff
@@ -0,0 +1,203 @@
+  prepareGame();
+}
+
+function prepareGame() {
```
Your function calls `prepareGame` but inside this function, you describe all game logic.
Single responsibility - your functions should do only one job. As an example function in which you fetch users should only fetch them and not render, transform or process them in other ways. All these things should be done in another place, outside your function. The same applied to the sort function, which usually does all types of sorting 😉

A-Ostrovnyy reviewed (changes requested):
Good progress!

Tedzury commented:
Good day. Applied target.closest to define clicked card. Also refactored prepare game function, splited it to func, that renders gameboard, func, that handles card click and resposible for game logic, and separetely pulled out rendering of scoreboard. Now it's rendered only once and then changing only one number of score. Previosly scoreboard was deleted everytime game resets.
